mlflow/tracing/utils/__init__.py

```diff
@@ -73,13 +73,40 @@ def default(self, obj):
         # representation, so we use dict representation instead.
         # E.g. https://github.com/run-llama/llama_index/blob/29ece9b058f6b9a1cf29bc723ed4aa3a39879ad5/llama-index-core/llama_index/core/chat_engine/types.py#L63-L64
         if is_dataclass(obj):
-            return asdict(obj)
+            try:
+                return asdict(obj)
+            except TypeError:
+                pass
+
+        # Some object has dangerous side effect in __str__ method, so we use class name instead.
+        if not self._is_safe_to_encode_str(obj):
+            return type(obj)
```

Comment:
```suggestion
            return type(obj).__name__
```

Is `__name__` missing? or json can handle this?

Reply:
Yes, supposedly it recursively encode the returned object from encoder.
```
 >>> json.dumps(StreamingAgentChatResponse(), cls=TraceJSONEncoder)
 '"<class \'llama_index.core.chat_engine.types.StreamingAgentChatResponse\'>"'
```